File: codes/visual_basic_dotnet/011_operator_spread.vb
Imports System.Collections.Generic ' Dictionary, KeyValuePair

Module Program
    Function JsonStringify(ByVal Anything As Object, Optional ByVal Pretty As Boolean = False, Optional ByVal Indent As String = "    ") As String
        Dim IndentLevel As Integer = 0
        Dim JsonStringifyInner As Func(Of Object, String, String) = Function(ByVal AnythingInner As Object, ByVal IndentInner As String)
            If (AnythingInner Is Nothing) Then Return "null"
            If (TypeOf AnythingInner Is String) Then Return """" & AnythingInner & """"
            If (IsNumeric(AnythingInner) = True) Then Return CStr(AnythingInner).Replace(",", ".")
            If (TypeOf AnythingInner Is Boolean) Then Return CStr(AnythingInner)
            If (TypeOf AnythingInner Is List(Of Object)) Then
                If (AnythingInner.Count = 0) Then Return "[]"
                IndentLevel += 1
                Dim Result As String = If((Pretty = True), "[" & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), "[")
                Dim ArrayItemIndex As Integer = 0
                For Each ArrayItem As Object In AnythingInner
                    Result &= JsonStringifyInner(ArrayItem, IndentInner)
                    If ((ArrayItemIndex + 1) <> AnythingInner.Count) Then Result &= If((Pretty = True), "," & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), ", ")
                    ArrayItemIndex += 1
                Next
                IndentLevel -= 1
                Result &= If((Pretty = True), Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) & "]", "]")
                Return Result
            End If
            If (TypeOf AnythingInner Is Dictionary(Of String, Object)) Then
                If (AnythingInner.Count = 0) Then Return "{}"
                IndentLevel += 1
                Dim Result As String = If((Pretty = True), "{" & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), "{")
                Dim ObjectIterationIndex As Integer = 0
                For Each ObjectEntry As KeyValuePair(Of String, Object) In AnythingInner
                    Dim ObjectKey = ObjectEntry.Key
                    Dim ObjectValue = ObjectEntry.Value
                    Result &= """" & ObjectKey & """: " & JsonStringifyInner(ObjectValue, IndentInner)
                    If ((ObjectIterationIndex + 1) <> AnythingInner.Count) Then Result &= If((Pretty = True), "," & Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)), ", ")
                    ObjectIterationIndex += 1
                Next
                IndentLevel -= 1
                Result &= If((Pretty = True), Environment.NewLine & String.Concat(Enumerable.Repeat(IndentInner, IndentLevel)) & "}", "}")
                Return Result
            End If
            Return "null"
        End Function
        Return JsonStringifyInner(Anything, Indent)
    End Function

    ' There's no JavaScript-like Spread Syntax (...) in Visual Basic (.NET).
    ' But, we can create our own function to mimic it in Visual Basic (.NET).

    Function SpreadObject(ByVal ParamArray Parameters() As Object) As Dictionary(Of String, Object)
        Dim NewObject As New Dictionary(Of String, Object)()
        Dim ParameterIndex As Integer = 0
        For Each Parameter As Object In Parameters
            If (TypeOf Parameter Is Dictionary(Of String, Object)) Then
                For Each ObjectEntry As KeyValuePair(Of String, Object) In Parameter
                    Dim ObjectKey = ObjectEntry.Key
                    Dim ObjectValue = ObjectEntry.Value
                    NewObject(CStr(ObjectKey)) = ObjectValue
                Next
                Continue For
            End If
            If (TypeOf Parameter Is List(Of Object)) Then
                Dim ArrayItemIndex As Integer = 0
                For Each ArrayItem As Object In Parameter
                    NewObject(CStr(ArrayItemIndex)) = ArrayItem
                    ArrayItemIndex += 1
                Next
                Continue For
            End If
            ParameterIndex += 1
        Next
        Return NewObject
    End Function

    Function SpreadArray(ByVal ParamArray Parameters() As Object) As List(Of Object)
        Dim NewArray As New List(Of Object)()
        Dim ParameterIndex As Integer = 0
        For Each Parameter As Object In Parameters
            If (TypeOf Parameter Is Dictionary(Of String, Object)) Then
                If (Parameter.Count = 1) Then
                    For Each ObjectEntry As KeyValuePair(Of String, Object) In Parameter
                        Dim ObjectKey = ObjectEntry.Key
                        Dim ObjectValue = ObjectEntry.Value
                        NewArray.Add(ObjectValue)
                    Next
                    Continue For
                End If
                NewArray.Add(Parameter)
                Continue For
            End If
            If (TypeOf Parameter Is List(Of Object)) Then
                Dim ArrayItemIndex As Integer = 0
                For Each ArrayItem As Object In Parameter
                    NewArray.Add(ArrayItem)
                    ArrayItemIndex += 1
                Next
                Continue For
            End If
            ParameterIndex += 1
        Next
        Return NewArray
    End Function

    Sub Main(Args As String())
        Console.WriteLine($"{Environment.NewLine}' JavaScript-like Spread Syntax (...) in Visual Basic (.NET)")

        Dim Fruits As List(Of Object) = New List(Of Object) From {"Mango", "Melon", "Banana"}
        Console.WriteLine("Fruits: " & JsonStringify(Fruits))

        Dim Vegetables As List(Of Object) = New List(Of Object) From {"Carrot", "Tomato"}
        Console.WriteLine("Vegetables: " & JsonStringify(Vegetables))

        Dim CountryCapitalsInAsia As Dictionary(Of String, Object) = New Dictionary(Of String, Object) From {
            {"Thailand", "Bangkok"},
            {"China", "Beijing"},
            {"Japan", "Tokyo"}
        }
        Console.WriteLine("CountryCapitalsInAsia: " & JsonStringify(CountryCapitalsInAsia, Pretty:=True))

        Dim CountryCapitalsInEurope As Dictionary(Of String, Object) = New Dictionary(Of String, Object) From {
            {"France", "Paris"},
            {"England", "London"}
        }
        Console.WriteLine("CountryCapitalsInEurope: " & JsonStringify(CountryCapitalsInEurope, Pretty:=True))

        Console.WriteLine($"{Environment.NewLine}' [...array1, ...array2]:{Environment.NewLine}")

        Dim Combination1 = SpreadArray(Fruits, Vegetables)
        Console.WriteLine("Combination1: " & JsonStringify(Combination1, Pretty:=True))
        ' Combination1: [
        '     "Mango",
        '     "Melon",
        '     "Banana",
        '     "Carrot",
        '     "Tomato"
        ' ]

        Dim Combination2 = SpreadArray(Fruits, New List(Of Object) From {"Cucumber", "Cabbage"})
        Console.WriteLine("Combination2: " & JsonStringify(Combination2, Pretty:=True))
        ' Combination2: [
        '     "Mango",
        '     "Melon",
        '     "Banana",
        '     "Cucumber",
        '     "Cabbage"
        ' ]

        Console.WriteLine($"{Environment.NewLine}' {{ ...object1, ...object2 }}:{Environment.NewLine}")

        Dim Combination3 = SpreadObject(CountryCapitalsInAsia, CountryCapitalsInEurope)
        Console.WriteLine("Combination3: " & JsonStringify(Combination3, Pretty:=True))
        ' Combination3: {
        '     "Thailand": "Bangkok",
        '     "China": "Beijing",
        '     "Japan": "Tokyo",
        '     "France": "Paris",
        '     "England": "London"
        ' }

        Dim Combination4 = SpreadObject(CountryCapitalsInAsia, New Dictionary(Of String, Object) From {{"Germany", "Berlin"}, {"Italy", "Rome"}})
        Console.WriteLine("Combination4: " & JsonStringify(Combination4, Pretty:=True))
        ' Combination4: {
        '     "Thailand": "Bangkok",
        '     "China": "Beijing",
        '     "Japan": "Tokyo",
        '     "Germany": "Berlin",
        '     "Italy": "Rome"
        ' }

        Console.WriteLine($"{Environment.NewLine}' [...array1, array2] || [...array1, newArrayItem1, newArrayItem2]:{Environment.NewLine}")

        Dim Combination5 = SpreadArray(Fruits, New Dictionary(Of String, Object) From {{"Vegetables", Vegetables}})
        Console.WriteLine("Combination5: " & JsonStringify(Combination5, Pretty:=True))
        ' Combination5: [
        '     "Mango",
        '     "Melon",
        '     "Banana",
        '     [
        '         "Carrot",
        '         "Tomato"
        '     ]
        ' ]

        Dim Combination6 = SpreadArray(Fruits, New Dictionary(Of String, Object) From {{"Vegetables", New List(Of Object) From {"Cucumber", "Cabbage"}}})
        Console.WriteLine("Combination6: " & JsonStringify(Combination6, Pretty:=True))
        ' Combination6: [
        '     "Mango",
        '     "Melon",
        '     "Banana",
        '     [
        '         "Cucumber",
        '         "Cabbage"
        '     ]
        ' ]

        Console.WriteLine($"{Environment.NewLine}' [...array1, object1] || [...array1, newArrayItem1, newArrayItem2]:{Environment.NewLine}")

        Dim Combination7 = SpreadArray(Fruits, New Dictionary(Of String, Object) From {{"CountryCapitalsInAsia", CountryCapitalsInAsia}})
        Console.WriteLine("Combination7: " & JsonStringify(Combination7, Pretty:=True))
        ' Combination7: [
        '     "Mango",
        '     "Melon",
        '     "Banana",
        '     {
        '         "Thailand": "Bangkok",
        '         "China": "Beijing",
        '         "Japan": "Tokyo"
        '     }
        ' ]

        Dim Combination8 = SpreadArray(Fruits, New Dictionary(Of String, Object) From {{"CountryCapitalsInEurope", New Dictionary(Of String, Object) From {{"Germany", "Berlin"}, {"Italy", "Rome"}}}})
        Console.WriteLine("Combination8: " & JsonStringify(Combination8, Pretty:=True))
        ' Combination8: [
        '     "Mango",
        '     "Melon",
        '     "Banana",
        '     {
        '         "Germany": "Berlin",
        '         "Italy": "Rome"
        '     }
        ' ]

        Console.WriteLine($"{Environment.NewLine}' {{ ...object1, object2 }} || {{ ...object1, objectKey: objectValue }}:{Environment.NewLine}")

        Dim Combination9 = SpreadObject(CountryCapitalsInAsia, New Dictionary(Of String, Object) From {{"CountryCapitalsInEurope", CountryCapitalsInEurope}})
        Console.WriteLine("Combination9: " & JsonStringify(Combination9, Pretty:=True))
        ' Combination9: {
        '     "Thailand": "Bangkok",
        '     "China": "Beijing",
        '     "Japan": "Tokyo",
        '     "CountryCapitalsInEurope": {
        '         "France": "Paris",
        '         "England": "London"
        '     }
        ' }

        Dim Combination10 = SpreadObject(CountryCapitalsInAsia, New Dictionary(Of String, Object) From {{"CountryCapitalsInEurope", New Dictionary(Of String, Object) From {{"Germany", "Berlin"}, {"Italy", "Rome"}}}})
        Console.WriteLine("Combination10: " & JsonStringify(Combination10, Pretty:=True))
        ' Combination10: {
        '     "Thailand": "Bangkok",
        '     "China": "Beijing",
        '     "Japan": "Tokyo",
        '     "CountryCapitalsInEurope": {
        '         "Germany": "Berlin",
        '         "Italy": "Rome"
        '     }
        ' }

        Console.WriteLine($"{Environment.NewLine}' {{ ...object1, array2 }} || {{ ...object1, objectKey: objectValue }}:{Environment.NewLine}")

        Dim Combination11 = SpreadObject(CountryCapitalsInAsia, New Dictionary(Of String, Object) From {{"Vegetables", Vegetables}})
        Console.WriteLine("Combination11: " & JsonStringify(Combination11, Pretty:=True))
        ' Combination11: {
        '     "Thailand": "Bangkok",
        '     "China": "Beijing",
        '     "Japan": "Tokyo",
        '     "Vegetables": [
        '         "Carrot",
        '         "Tomato"
        '     ]
        ' }

        Dim Combination12 = SpreadObject(CountryCapitalsInAsia, New Dictionary(Of String, Object) From {{"Vegetables", New List(Of Object) From {"Cucumber", "Cabbage"}}})
        Console.WriteLine("Combination12: " & JsonStringify(Combination12, Pretty:=True))
        ' Combination12: {
        '     "Thailand": "Bangkok",
        '     "China": "Beijing",
        '     "Japan": "Tokyo",
        '     "Vegetables": [
        '         "Cucumber",
        '         "Cabbage"
        '     ]
        ' }

        Console.WriteLine($"{Environment.NewLine}' {{ ...object1, ...array2 }}:{Environment.NewLine}")

        Dim Combination13 = SpreadObject(CountryCapitalsInAsia, Vegetables)
        Console.WriteLine("Combination13: " & JsonStringify(Combination13, Pretty:=True))
        ' Combination13: {
        '     "Thailand": "Bangkok",
        '     "China": "Beijing",
        '     "Japan": "Tokyo",
        '     "0": "Carrot",
        '     "1": "Tomato"
        ' }

        Dim Combination14 = SpreadObject(CountryCapitalsInAsia, New List(Of Object) From {"Cucumber", "Cabbage"})
        Console.WriteLine("Combination14: " & JsonStringify(Combination14, Pretty:=True))
        ' Combination14: {
        '     "Thailand": "Bangkok",
        '     "China": "Beijing",
        '     "Japan": "Tokyo",
        '     "0": "Cucumber",
        '     "1": "Cabbage"
        ' }

        ' Console.WriteLine($"{Environment.NewLine}' [...array1, ...object1]: ` this combination throw an error in JavaScript:{Environment.NewLine}")

        ' this combination throw an error in JavaScript
        ' Dim CombinationErrorInJavascript1 = SpreadArray(Fruits, CountryCapitalsInAsia)
        ' Console.WriteLine("CombinationErrorInJavascript1: " & JsonStringify(CombinationErrorInJavascript1, Pretty:=True))

        ' this combination throw an error in JavaScript
        ' Dim CombinationErrorInJavascript2 = SpreadArray(Fruits, New Dictionary(Of String, Object) From {{"Germany", "Berlin"}, {"Italy", "Rome"}})
        ' Console.WriteLine("CombinationErrorInJavascript2: " & JsonStringify(CombinationErrorInJavascript2, Pretty:=True))
    End Sub
End Module

File: codes/gnu_octave/011_operator_spread.m
function stringrepeatresult = stringrepeat(astring, count)
    result = "";
    for i = (1:1:count) % (start:step:stop)
        result = cstrcat(result, astring);
    end
    stringrepeatresult = result;
end

function arrayreduceresult = arrayreduce(callbackfunction, anarray, initialvalue)
    % JavaScript-like Array.reduce() function
    currentresult = initialvalue;
    for arrayitemindex = (1:1:numel(anarray)) % (start:step:stop)
        arrayitem = anarray{arrayitemindex};
        currentresult = callbackfunction(currentresult, arrayitem, arrayitemindex, anarray);
    end
    arrayreduceresult = currentresult;
end

function optionalchainingresult = optionalchaining(anything, varargin)
    % JavaScript-like Optional Chaining Operator (?.) function
    % varargin == objectpropertiesarray
    if ((~isstruct(anything) && ~iscell(anything)) || isempty(varargin))
        optionalchainingresult = anything;
        return;
    end
    function arrayreducecallbackresult = arrayreducecallback(currentresult, currentitem, varargin)
        % varargin == (arrayitemindex, anarray)
        if ((isempty(currentresult) && isstruct(anything) && (ischar(currentitem) || ischar(currentitem))) && isfield(anything, currentitem))
            arrayreducecallbackresult = anything.(currentitem);
            return;
        end
        if (isempty(currentresult) && iscell(anything) && isnumeric(currentitem) && (currentitem >= 1) && (numel(anything) > currentitem))
            arrayreducecallbackresult = anything{currentitem};
            return;
        end
        if ((isstruct(currentresult) && (ischar(currentitem) || ischar(currentitem))) && isfield(currentresult, currentitem))
            arrayreducecallbackresult = currentresult.(currentitem);
            return;
        end
        if (iscell(currentresult) && isnumeric(currentitem) && (currentitem >= 1) && (numel(currentresult) > currentitem))
            arrayreducecallbackresult = currentresult{currentitem};
            return;
        end
        arrayreducecallbackresult = {};
    end
    optionalchainingresult = arrayreduce(@arrayreducecallback, varargin, {});
end

function ternaryresult = ternary(truecondition, callbackfunctionifconditiontrue, callbackfunctionifconditionfalse)
    if (islogical(truecondition) && (truecondition == true))
        ternaryresult = callbackfunctionifconditiontrue();
        return;
    end
    ternaryresult = callbackfunctionifconditionfalse();
end

function jsonstringifyresult = jsonstringify(anything, varargin)
    indentlevel = 0;
    function jsonstringifyinnerresult = jsonstringifyinner(anythinginner, prettyinner, indentinner)
        if isempty(anythinginner)
            jsonstringifyinnerresult = "null";
            return;
        end
        if (ischar(anythinginner) || ischar(anythinginner))
            jsonstringifyinnerresult = cstrcat("""", anythinginner, """");
            return;
        end
        if isnumeric(anythinginner)
            jsonstringifyinnerresult = num2str(anythinginner);
            return;
        end
        if (islogical(anythinginner) && (anythinginner == true))
            jsonstringifyinnerresult = "true";
            return;
        end
        if (islogical(anythinginner) && (anythinginner == false))
            jsonstringifyinnerresult = "false";
            return;
        end
        if iscell(anythinginner)
            if isempty(anythinginner)
                jsonstringifyinnerresult = "[]";
                return;
            end
            indentlevel = indentlevel + 1;
            result = ternary((pretty == true), @() (cstrcat("[", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() ("["));
            for arrayitemindex = (1:1:numel(anythinginner)) % (start:step:stop)
                arrayitem = anythinginner{arrayitemindex};
                result = cstrcat(result, jsonstringifyinner(arrayitem, prettyinner, indentinner));
                if (arrayitemindex ~= numel(anythinginner))
                    result = ternary((pretty == true), @() (cstrcat(result, ",", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() (cstrcat(result, ", ")));
                end
            end
            indentlevel = indentlevel - 1;
            result = ternary((pretty == true), @() (cstrcat(result, sprintf("\n"), stringrepeat(indentinner, indentlevel), "]")), @() (cstrcat(result, "]")));
            jsonstringifyinnerresult = result;
            return;
        end
        if isstruct(anythinginner)
            if (numel(fieldnames(anythinginner)) == 0)
                jsonstringifyinnerresult = "{}";
                return;
            end
            indentlevel = indentlevel + 1;
            result = ternary((pretty == true), @() (cstrcat("{", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() ("{"));
            objectkeys = fieldnames(anythinginner);
            for objectentryindex = (1:1:numel(objectkeys)) % (start:step:stop)
                objectkey = objectkeys{objectentryindex};
                objectvalue = anythinginner.(objectkey);
                result = cstrcat(result, """", objectkey, """: ", jsonstringifyinner(objectvalue, prettyinner, indentinner));
                if (objectentryindex ~= numel(objectkeys))
                    result = ternary((pretty == true), @() (cstrcat(result, ",", sprintf("\n"), stringrepeat(indentinner, indentlevel))), @() (cstrcat(result, ", ")));
                end
            end
            indentlevel = indentlevel - 1;
            result = ternary((pretty == true), @() (cstrcat(result, sprintf("\n"), stringrepeat(indentinner, indentlevel), "}")), @() (cstrcat(result, "}")));
            jsonstringifyinnerresult = result;
            return;
        end
        jsonstringifyinnerresult = "null";
    end
    prettydefault = false;
    indentdefault = "    ";
    pretty = prettydefault;
    indent = indentdefault;
    if isempty(varargin)
        jsonstringifyresult = jsonstringifyinner(anything, prettydefault, indentdefault);
        return;
    end
    optionalargument = varargin{1};
    if (islogical(optionalargument) && (optionalargument == false))
        jsonstringifyresult = jsonstringifyinner(anything, prettydefault, indentdefault);
        return;
    end
    if isstruct(optionalargument)
        pretty = optionalchaining(optionalargument, "pretty");
        indent = optionalchaining(optionalargument, "indent");
        pretty = ternary(isempty(pretty), @() (prettydefault), @() (pretty));
        indent = ternary(isempty(indent), @() (indentdefault), @() (indent));
        jsonstringifyresult = jsonstringifyinner(anything, pretty, indent);
        return;
    end
    if (islogical(optionalargument) && (optionalargument == true))
        if (numel(varargin) >= 2)
            optionalargument2 = varargin{2};
            if (ischar(optionalargument2) || ischar(optionalargument2))
                indent = optionalargument2
            end
        end
        pretty = optionalargument;
        jsonstringifyresult = jsonstringifyinner(anything, pretty, indent);
        return;
    end
    jsonstringifyresult = jsonstringifyinner(anything, prettydefault, indentdefault);
end

function sprint(varargin)
    currentresult = "";
    for argumentindex = (1:1:numel(varargin)) % (start:step:stop)
        argument = varargin{argumentindex};
        if (iscell(argument) && (numel(argument) == 1))
            argumentnew = jsonstringify(argument{1});
            argumentnew = strrep(argumentnew, """{", "{");
            argumentnew = strrep(argumentnew, """[", "[");
            argumentnew = strrep(argumentnew, "}""", "}");
            argumentnew = strrep(argumentnew, "]""", "]");
            currentresult = cstrcat(currentresult, argumentnew);
            continue;
        end
        if (~ischar(argument) && ~ischar(argument))
            error("Non string argument must be wrapped in {}");
            continue;
        end
        currentresult = cstrcat(currentresult, argument);
    end
    disp(currentresult);
end

sprint(sprintf("\n"), "% JavaScript-like Spread Syntax (...) in GNU Octave");

% There's no JavaScript-like Spread Syntax (...) in GNU Octave.
% But, we can create our own function to mimic it in GNU Octave.

function spreadobjectresult = spreadobject(varargin)
    newobject = struct();
    for argumentindex = (1:1:numel(varargin)) % (start:step:stop)
        argument = varargin{argumentindex};
        if isstruct(argument)
            objectkeys = fieldnames(argument);
            for objectentryindex = (1:1:numel(objectkeys)) % (start:step:stop)
                objectkey = objectkeys{objectentryindex};
                objectvalue = argument.(objectkey);
                newobject.(objectkey) = objectvalue;
            end
            continue;
        end
        if iscell(argument)
            for arrayitemindex = (1:1:numel(argument)) % (start:step:stop)
                arrayitem = argument{arrayitemindex};
                newobject.(cstrcat("index", num2str(arrayitemindex))) = arrayitem;
            end
            continue;
        end
    end
    spreadobjectresult = newobject;
end

function spreadarrayresult = spreadarray(varargin)
    newarray = {};
    for argumentindex = (1:1:numel(varargin)) % (start:step:stop)
        argument = varargin{argumentindex};
        if isstruct(argument)
            objectkeys = fieldnames(argument);
            if (numel(objectkeys) == 1)
                for objectentryindex = (1:1:numel(objectkeys)) % (start:step:stop)
                    objectkey = objectkeys{objectentryindex};
                    objectvalue = argument.(objectkey);
                    newarray{end + 1} = objectvalue;
                end
                continue;
            end
            newarray{end + 1} = argument;
            continue;
        end
        if iscell(argument)
            for arrayitemindex = (1:1:numel(argument)) % (start:step:stop)
                arrayitem = argument{arrayitemindex};
                newarray{end + 1} = arrayitem;
            end
            continue;
        end
    end
    spreadarrayresult = newarray;
end

fruits = {"Mango", "Melon", "Banana"};
sprint("fruits: ", jsonstringify(fruits));

vegetables = {"Carrot", "Tomato"};
sprint("vegetables: ", jsonstringify(vegetables));

countrycapitalsinasia = struct( ...
    "Thailand", {"Bangkok"}, ...
    "China", {"Beijing"}, ...
    "Japan", {"Tokyo"} ...
);
sprint("countrycapitalsinasia: ", jsonstringify(countrycapitalsinasia, struct("pretty", {true})));

countrycapitalsineurope = struct( ...
    "France", {"Paris"}, ...
    "England", {"London"} ...
);
sprint("countrycapitalsineurope: ", jsonstringify(countrycapitalsineurope, struct("pretty", {true})));

sprint(sprintf("\n"), "% [...array1, ...array2]:", sprintf("\n"));

combination1 = spreadarray(fruits, vegetables);
sprint("combination1: ", jsonstringify(combination1, struct("pretty", {true})));
% combination1: [
%     "Mango",
%     "Melon",
%     "Banana",
%     "Carrot",
%     "Tomato"
% ]

combination2 = spreadarray(fruits, {"Cucumber", "Cabbage"});
sprint("combination2: ", jsonstringify(combination2, struct("pretty", {true})));
% combination2: [
%     "Mango",
%     "Melon",
%     "Banana",
%     "Cucumber",
%     "Cabbage"
% ]

sprint(sprintf("\n"), "% { ...object1, ...object2 }:", sprintf("\n"));

combination3 = spreadobject(countrycapitalsinasia, countrycapitalsineurope);
sprint("combination3: ", jsonstringify(combination3, struct("pretty", {true})));
% combination3: {
%     "Thailand": "Bangkok",
%     "China": "Beijing",
%     "Japan": "Tokyo",
%     "France": "Paris",
%     "England": "London"
% }

combination4 = spreadobject(countrycapitalsinasia, struct("Germany", {"Berlin"}, "Italy", {"Rome"}));
sprint("combination4: ", jsonstringify(combination4, struct("pretty", {true})));
% combination4: {
%     "Thailand": "Bangkok",
%     "China": "Beijing",
%     "Japan": "Tokyo",
%     "Germany": "Berlin",
%     "Italy": "Rome"
% }

sprint(sprintf("\n"), "% [...array1, array2] || [...array1, newArrayItem1, newArrayItem2]:", sprintf("\n"));

combination5 = spreadarray(fruits, struct("vegetables", {vegetables}));
sprint("combination5: ", jsonstringify(combination5, struct("pretty", {true})));
% combination5: [
%     "Mango",
%     "Melon",
%     "Banana",
%     [
%         "Carrot",
%         "Tomato"
%     ]
% ]

combination6 = spreadarray(fruits, struct("vegetables", {{"Cucumber", "Cabbage"}}));
sprint("combination6: ", jsonstringify(combination6, struct("pretty", {true})));
% combination6: [
%     "Mango",
%     "Melon",
%     "Banana",
%     [
%         "Cucumber",
%         "Cabbage"
%     ]
% ]

sprint(sprintf("\n"), "% [...array1, object1] || [...array1, newArrayItem1, newArrayItem2]:", sprintf("\n"));

combination7 = spreadarray(fruits, struct("countrycapitalsinasia", {countrycapitalsinasia}));
sprint("combination7: ", jsonstringify(combination7, struct("pretty", {true})));
% combination7: [
%     "Mango",
%     "Melon",
%     "Banana",
%     {
%         "Thailand": "Bangkok",
%         "China": "Beijing",
%         "Japan": "Tokyo"
%     }
% ]

combination8 = spreadarray(fruits, struct("countrycapitalsineurope", {struct("Germany", {"Berlin"}, "Italy", {"Rome"})}));
sprint("combination8: ", jsonstringify(combination8, struct("pretty", {true})));
% combination8: [
%     "Mango",
%     "Melon",
%     "Banana",
%     {
%         "Germany": "Berlin",
%         "Italy": "Rome"
%     }
% ]

sprint(sprintf("\n"), "% { ...object1, object2 } || { ...object1, objectKey: objectValue }:", sprintf("\n"));

combination9 = spreadobject(countrycapitalsinasia, struct("countrycapitalsineurope", {countrycapitalsineurope}));
sprint("combination9: ", jsonstringify(combination9, struct("pretty", {true})));
% combination9: {
%     "Thailand": "Bangkok",
%     "China": "Beijing",
%     "Japan": "Tokyo",
%     "countrycapitalsineurope": {
%         "France": "Paris",
%         "England": "London"
%     }
% }

combination10 = spreadobject(countrycapitalsinasia, struct("countrycapitalsineurope", {struct("Germany", {"Berlin"}, "Italy", {"Rome"})}));
sprint("combination10: ", jsonstringify(combination10, struct("pretty", {true})));
% combination10: {
%     "Thailand": "Bangkok",
%     "China": "Beijing",
%     "Japan": "Tokyo",
%     "countrycapitalsineurope": {
%         "Germany": "Berlin",
%         "Italy": "Rome"
%     }
% }

sprint(sprintf("\n"), "% { ...object1, array2 } || { ...object1, objectKey: objectValue }:", sprintf("\n"));

combination11 = spreadobject(countrycapitalsinasia, struct("vegetables", {vegetables}));
sprint("combination11: ", jsonstringify(combination11, struct("pretty", {true})));
% combination11: {
%     "Thailand": "Bangkok",
%     "China": "Beijing",
%     "Japan": "Tokyo",
%     "vegetables": [
%         "Carrot",
%         "Tomato"
%     ]
% }

combination12 = spreadobject(countrycapitalsinasia, struct("vegetables", {{"Cucumber", "Cabbage"}}));
sprint("combination12: ", jsonstringify(combination12, struct("pretty", {true})));
% combination12: {
%     "Thailand": "Bangkok",
%     "China": "Beijing",
%     "Japan": "Tokyo",
%     "vegetables": [
%         "Cucumber",
%         "Cabbage"
%     ]
% }

sprint(sprintf("\n"), "% { ...object1, ...array2 }:", sprintf("\n"));

combination13 = spreadobject(countrycapitalsinasia, vegetables);
sprint("combination13: ", jsonstringify(combination13, struct("pretty", {true})));
% combination13: {
%     "Thailand": "Bangkok",
%     "China": "Beijing",
%     "Japan": "Tokyo",
%     "index1": "Carrot",
%     "index2": "Tomato"
% }

combination14 = spreadobject(countrycapitalsinasia, {"Cucumber", "Cabbage"});
sprint("combination14: ", jsonstringify(combination14, struct("pretty", {true})));
% combination14: {
%     "Thailand": "Bangkok",
%     "China": "Beijing",
%     "Japan": "Tokyo",
%     "index1": "Carrot",
%     "index2": "Tomato"
% }

% sprint(sprintf("\n"), "% [...array1, ...object1]: % this combination throw an error in JavaScript", sprintf("\n"));

% this combination throw an error in JavaScript
% combinationerrorinjavascript1 = spreadarray(fruits, countrycapitalsinasia);
% sprint("combinationerrorinjavascript1: ", jsonstringify(combinationerrorinjavascript1, struct("pretty", {true})));

% this combination throw an error in JavaScript
% combinationerrorinjavascript2 = spreadarray(fruits, struct("Germany", {"Berlin"}, "Italy", {"Rome"}));
% sprint("combinationerrorinjavascript2: ", jsonstringify(combinationerrorinjavascript2, struct("pretty", {true})));
